keycloak-rest-api-extensions/src/main/java/io/cloudtrust/keycloak/authentication/actiontoken/CtExecuteActionsActionTokenHandler.java

```diff
@@ -0,0 +1,139 @@
+        final UriInfo uriInfo = tokenContext.getUriInfo();
+        final RealmModel realm = tokenContext.getRealm();
+        final KeycloakSession session = tokenContext.getSession();
+        if (tokenContext.isAuthenticationSessionFresh()) {
+            // Update the authentication session in the token
+            String authSessionEncodedId = AuthenticationSessionCompoundId.fromAuthSession(authSession).getEncodedId();
+            token.setCompoundAuthenticationSessionId(authSessionEncodedId);
+            UriBuilder builder = Urls.actionTokenBuilder(uriInfo.getBaseUri(), token.serialize(session, realm, uriInfo),
+                    authSession.getClient().getClientId(), authSession.getTabId());
+            String confirmUri = builder.build(realm.getName()).toString();
+
+            return session.getProvider(LoginFormsProvider.class)
+                    .setAuthenticationSession(authSession)
+                    .setSuccess(Messages.CONFIRM_EXECUTION_OF_ACTIONS)
+                    .setAttribute(Constants.TEMPLATE_ATTR_ACTION_URI, confirmUri)
+                    .setAttribute(Constants.TEMPLATE_ATTR_REQUIRED_ACTIONS, token.getRequiredActions())
+                    .createInfoPage();
+        }
+
+        String redirectUri = RedirectUtils.verifyRedirectUri(tokenContext.getSession(), token.getRedirectUri(), authSession.getClient());
+        if (redirectUri != null) {
+            authSession.setAuthNote(AuthenticationManager.SET_REDIRECT_URI_AFTER_REQUIRED_ACTIONS, "true");
+
+            authSession.setRedirectUri(redirectUri);
+            authSession.setClientNote(OIDCLoginProtocol.REDIRECT_URI_PARAM, redirectUri);
+        }
+
+        token.getRequiredActions().stream().forEach(authSession::addRequiredAction);
+
+        UserModel user = tokenContext.getAuthenticationSession().getAuthenticatedUser();
+        // verify user email as we know it is valid as this entry point would never have gotten here.
+        String emailClaim = (String) token.getOtherClaims().get(CtExecuteActionsActionToken.CLAIM_EMAIL);
+        setEmailVerified(user, emailClaim);
+
+        String nextAction = AuthenticationManager.nextRequiredAction(tokenContext.getSession(), authSession, tokenContext.getRequest(), tokenContext.getEvent());
+        return AuthenticationManager.redirectToRequiredActions(tokenContext.getSession(), tokenContext.getRealm(), authSession, tokenContext.getUriInfo(), nextAction);
+    }
+
+    private void setEmailVerified(UserModel user, String emailClaim) {
+        if (StringUtils.isBlank(emailClaim)) {
+            // Email to validate is not present in the token so we are basically checking the main email (user::getEmail)
+            // If we reached this point, the email is valid
+            user.setEmailVerified(true);
+            return;
+        }
+        if (!emailClaim.equalsIgnoreCase(user.getFirstAttribute(UserUtils.ATTRB_EMAIL_TO_VALIDATE))) {
+            return;
```

Comment:
End user might think there is a bug as no error will be displayed ?
Maybe we should throw an error that state the request is no more valid ?